drivers/rtc/rtc_numaker.c

```diff
@@ -0,0 +1,401 @@
+	uint32_t hour;		/* Hour value */
+	uint32_t minute;	/* Minute value */
+	uint32_t second;	/* Second value */
+	uint32_t time_scale;	/* 12-Hour, 24-Hour */
+	uint32_t am_pm;		/* Only Time Scale select 12-hr used */
+};
+
+static int rtc_numaker_set_time(const struct device *dev, const struct rtc_time *timeptr)
+{
+	struct rtc_numaker_time curr_time;
+	struct rtc_numaker_data *data = dev->data;
+	uint32_t real_year = timeptr->tm_year + TM_YEAR_REF;
+#ifdef CONFIG_RTC_ALARM
+	const struct rtc_numaker_config *config = dev->config;
+	RTC_T *rtc_base = config->rtc_base;
+#endif
+
+	if (real_year < NVT_RTC_YEAR_MIN || real_year > NVT_RTC_YEAR_MAX) {
+		/* RTC can't support years out of 2000 ~ 2099 */
+		return -EINVAL;
+	}
+
+	if (timeptr->tm_wday == -1) {
+		return -EINVAL;
+	}
+
+	k_spinlock_key_t key = k_spin_lock(&data->lock);
+
+	curr_time.year = real_year;
+	curr_time.month = timeptr->tm_mon + 1;
+	curr_time.day = timeptr->tm_mday;
+	curr_time.hour = timeptr->tm_hour;
+	curr_time.minute = timeptr->tm_min;
+	curr_time.second = timeptr->tm_sec;
+	curr_time.day_of_week = timeptr->tm_wday;
+	curr_time.time_scale = NVT_TIME_SCALE;
+
+	RTC_SetDateAndTime((S_RTC_TIME_DATA_T *)&curr_time);
+
+#ifdef CONFIG_RTC_ALARM
+	/* Enable RTC alarm mask */
+	rtc_base->CAMSK = data->camsk;
+	rtc_base->TAMSK = data->tamsk;
+#endif
+	k_spin_unlock(&data->lock, key);
+
+	return 0;
+}
+
+static int rtc_numaker_get_time(const struct device *dev, struct rtc_time *timeptr)
+{
+	struct rtc_numaker_data *data = dev->data;
+	struct rtc_numaker_time curr_time;
+	k_spinlock_key_t key = k_spin_lock(&data->lock);
+
+	curr_time.time_scale = NVT_TIME_SCALE;
+	RTC_GetDateAndTime((S_RTC_TIME_DATA_T *)&curr_time);
+	k_spin_unlock(&data->lock, key);
+
+	timeptr->tm_year = curr_time.year - TM_YEAR_REF;
+	timeptr->tm_mon = curr_time.month - 1;
+	timeptr->tm_mday = curr_time.day;
+	timeptr->tm_wday = curr_time.day_of_week;
+
+	timeptr->tm_hour = curr_time.hour;
+	timeptr->tm_min = curr_time.minute;
+	timeptr->tm_sec = curr_time.second;
+	timeptr->tm_nsec = 0;
+
+	/* unknown values */
+	timeptr->tm_yday = -1;
+	timeptr->tm_isdst = -1;
+
+	return 0;
+}
+
+static void rtc_numaker_isr(const struct device *dev)
+{
+	const struct rtc_numaker_config *config = dev->config;
+	RTC_T *rtc_base = config->rtc_base;
+	uint32_t int_status;
+#ifdef CONFIG_RTC_ALARM
+	struct rtc_numaker_data *data = dev->data;
+#endif
+
+	int_status = rtc_base->INTSTS;
+	if (int_status & RTC_INTSTS_TICKIF_Msk) {
+		/* Clear RTC Tick interrupt flag */
+		rtc_base->INTSTS = RTC_INTSTS_TICKIF_Msk;
+	}
+
+#ifdef CONFIG_RTC_ALARM
+	if (int_status & RTC_INTSTS_ALMIF_Msk) {
+		rtc_alarm_callback callback;
+		void *user_data;
+
+		/* Clear RTC Alarm interrupt flag */
+		rtc_base->INTSTS = RTC_INTSTS_ALMIF_Msk;
+		rtc_base->CAMSK = 0x00;
+		rtc_base->TAMSK = 0x00;
+		K_SPINLOCK(&data->lock) {
+			callback = data->alarm_callback;
+			user_data = data->alarm_user_data;
+			data->alarm_pending = callback ? false : true;
+		}
```

Comment:
Instead of taking a spinlock in the ISR to avoid data races, disable the ISR after taking the spinlock in the API handlers. This solution results in the following example flow:
1. `rtc_numaker_alarm_set_time` is entered
2. spin lock is taken at line 214
3. now, we can disable the RTC ISR
4. We can now safely modify data shared between isr and calling context
5. reenable RTC ISR
6. release spinlock

Reply:
it' done to avoid taking spinlock in ISR by disable ISR in API handler.
More, it could pass rtc_api test except H/W not support year < 2000.
```
------ TESTSUITE SUMMARY START ------

SUITE PASS - 100.00% [rtc_api]: pass = 4, fail = 0, skip = 1, total = 5 duration = 19.059 seconds
 - PASS - [rtc_api.test_alarm] duration = 4.526 seconds
 - PASS - [rtc_api.test_alarm_callback] duration = 4.526 seconds
 - PASS - [rtc_api.test_set_get_time] duration = 0.001 seconds
 - PASS - [rtc_api.test_time_counting] duration = 10.003 seconds
 - SKIP - [rtc_api.test_y2k] duration = 0.003 seconds

------ TESTSUITE SUMMARY END ------

===================================================================
PROJECT EXECUTION SUCCESSFUL
```